_overviews/core/architecture-of-scala-collections.md

```diff
@@ -631,144 +684,195 @@ So the result of `map` and `++` is never an RNA strand, even if the
+`prependedAll`, `map` and `flatMap`.
 
-      private val S = 2            // number of bits in group
-      private val M = (1 << S) - 1 // bitmask to isolate a group
-      private val N = 32 / S       // number of groups in an Int
+This implementation now behaves correctly, but we can still improve a few things. Since our
+collection is strict, we could take advantage of strict builders in transformation operations.
+Also, if we try to convert an `Iterable[Base]` into an `RNA3` it fails:
 
-      def fromSeq(buf: Seq[Base]): RNA = {
-        val groups = new Array[Int]((buf.length + N - 1) / N)
-        for (i <- 0 until buf.length)
-          groups(i / N) |= Base.toInt(buf(i)) << (i % N * S)
-        new RNA(groups, buf.length)
-      }
+~~~
+scala> val bases: Iterable[Base] = List(A, U, C, C)
+bases: Iterable[Base] = List(A, U, C, C)
 
-      def apply(bases: Base*) = fromSeq(bases)
+scala> bases.to(RNA)
+                ^
+       error: type mismatch;
+        found   : scala.collection.RNA.type
+        required: scala.collection.Factory[strawman.collection.Base,?]
+~~~
 
-      def newBuilder: Builder[Base, RNA] =
-        new ArrayBuffer mapResult fromSeq
+### Final version of RNA strands class ###
 
-      implicit def canBuildFrom: CanBuildFrom[RNA, Base, RNA] =
-        new CanBuildFrom[RNA, Base, RNA] {
-          def apply(): Builder[Base, RNA] = newBuilder
-          def apply(from: RNA): Builder[Base, RNA] = newBuilder
-        }
+~~~ scala
+final class RNA private (
+  val groups: Array[Int],
+  val length: Int
+) extends IndexedSeq[Base]
+    with IndexedSeqOps[Base, IndexedSeq, RNA]
+    with StrictOptimizedSeqOps[Base, IndexedSeq, RNA] { rna =>
+
+  import RNA._
+
+  // Mandatory implementation of `apply` in `IndexedSeqOps`
+  def apply(idx: Int): Base = {
+    if (idx < 0 || length <= idx)
+      throw new IndexOutOfBoundsException
+    Base.fromInt(groups(idx / N) >> (idx % N * S) & M)
+  }
+
+  // Mandatory implementation of `iterableFactory`,
+  // `fromSpecificIterable` and `newSpecificBuilder`,
+  // from `IterableOps`
+  def iterableFactory: SeqFactory[IndexedSeq] = IndexedSeq
+  protected[this] def fromSpecificIterable(coll: Iterable[Base]): RNA =
+    RNA.fromSpecific(coll)
+  protected[this] def newSpecificBuilder(): Builder[Base, RNA] =
+    RNA.newBuilder()
+
+  // Overloading of `appended`, `prepended`, `appendedAll`, `prependedAll`,
+  // `map` and `flatMap` to return an `RNA` when possible
+  def appended(base: Base): RNA = 
+    (newSpecificBuilder() ++= this += base).result()
+  def appendedAll(suffix: Iterable[Base]): RNA =
+    (newSpecificBuilder() ++= this ++= suffix).result()
+  def prepended(base: Base): RNA = 
+    (newSpecificBuilder() += base ++= this).result()
+  def prependedAll(prefix: Iterable[Base]): RNA =
+    (newSpecificBuilder() ++= prefix ++= this).result()
+  def map(f: Base => Base): RNA = {
+    var b = newSpecificBuilder()
+    for (base <- this) {
+      b += f(base)
     }
```

Comment:
I was also wondering if we should provide helpers to implement such operations in a strict way. For the view based version it is as easy as writing:

~~~ scala
def map(f: Base => Base): RNA = fromSpecificIterable(View.Map(this, f))
~~~

We should probably have an equivalent strict version:

~~~ scala
object Builder {

  def map[CC[_], A, B](source: IterableOnce[A], f: A => B, builder: Builder[B, CC[B]]): CC[B] = {
    for (a <- source) {
      b += f(base)
    }
    builder.result()
  }
}
~~~

So that users could just write:

~~~ scala
def map(f: Base => Base): RNA = Builder.map(this, f, newSpecificBuilder())
~~~

And we could use it also in [`StrictOptimizedIterableOps`](https://github.com/scala/collection-strawman/blob/master/collections/src/main/scala/strawman/collection/StrictOptimizedIterableOps.scala#L60-L67).

The only part that I don’t like is that the builder is created in one place (user code) and used in a different place (framework code). This can be error prone because we really want the builder to be freshly created.